test/jdk/javax/swing/JButton/bug4323121.java

```diff
@@ -30,96 +30,100 @@
  * @run main bug4323121
  */
 
-import java.awt.Graphics;
 import java.awt.Point;
 import java.awt.Robot;
+import java.awt.event.MouseAdapter;
 import java.awt.event.MouseEvent;
-import java.awt.event.MouseListener;
-import java.awt.event.MouseMotionListener;
+import java.awt.event.WindowAdapter;
+import java.awt.event.WindowEvent;
+import java.util.concurrent.CountDownLatch;
+
 import javax.swing.JButton;
 import javax.swing.JFrame;
 import javax.swing.SwingUtilities;
 
-public class bug4323121 {
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+public final class bug4323121 {
 
     static JFrame frame;
-    static testButton button;
-    static volatile Point pt;
-    static volatile int buttonW;
-    static volatile int buttonH;
-    static volatile boolean failed = false;
+    static JButton button;
+
+    static volatile Point buttonCenter;
+
+    private static final CountDownLatch windowGainedFocus = new CountDownLatch(1);
+
+    private static final CountDownLatch mouseEntered = new CountDownLatch(1);
+
+    // Usage of this flag is thread-safe because of using the mouseEntered latch
+    private static boolean modelArmed;
 
     public static void main(String[] args) throws Exception {
         Robot robot = new Robot();
         robot.setAutoDelay(100);
+
         try {
             SwingUtilities.invokeAndWait(() -> {
+                button = new TestButton("gotcha");
+                button.addMouseListener(new MouseAdapter() {
+                    @Override
+                    public void mouseEntered(MouseEvent e) {
+                        if (button.getModel().isArmed()) {
+                            modelArmed = true;
+                        }
+                        mouseEntered.countDown();
+                    }
+                });
+
                 frame = new JFrame("bug4323121");
-                button = new testButton("gotcha");
                 frame.getContentPane().add(button);
+
+                frame.addWindowFocusListener(new WindowAdapter() {
+                    @Override
+                    public void windowGainedFocus(WindowEvent e) {
+                        windowGainedFocus.countDown();
+                    }
+                });
+
                 frame.pack();
                 frame.setLocationRelativeTo(null);
                 frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
                 frame.setVisible(true);
             });
+
+            if (!windowGainedFocus.await(1, SECONDS)) {
+                throw new RuntimeException("Window didn't gain focus");
+            }
             robot.waitForIdle();
-            robot.delay(1000);
```

Comment:
The new code probably should follow the same pattern, flush all events/commands to native by the waitForIdle. and then wait response from OS for 1_or_less second.

Reply:
I reversed the order because window may gain focus before all the events are processed.

To ensure all the events in the queue are handled, I added `robot.waitForIdle()`.

The next action in the test is to get location of the button, which can be done safely since the frame must be displaying on the screen because it has already received `windowGainedFocus` event.

For this reason, there's no need for `robot.delay(1000)`.

Reply:
This test follows the same pattern that I used in https://github.com/openjdk/jdk/pull/21474:

https://github.com/openjdk/jdk/pull/21474/files#diff-2a8cee50408407d7e24de4cb85f0921cfc5db7c215d3b29cf101a4050323e9a9R92

Reply:
> To ensure all the events in the queue are handled, I added `robot.waitForIdle()`.

This will flush events which were posted after the windows gained focus, but before the patch the test flushed all events (including event to show the window which might be asynchronous). My point is that this operation might be longer that 1 second(one of the reason we have all that deleyas). So we should waitForIdle first, then wait via delay or other sync_ops.

Reply:
This will flush all the pending events in the queue, which includes events that were posted before the `windowGainedFocus` latch is released as well as events are posted afterwards.

There's no difference whether I place `waitForIdle` before or after.

If I place `robot.waitForIdle` before `windowGainedFocus.await`, then `await` immediately returns, that is the delay added by `await` is zero.

If I place `robot.waitForIdle` after `windowGainedFocus.await`, the main thread stops until the window gained focus event is delivered, after which `robot.waitForIdle` flushes the remaining events in the queue.

In both cases, there's a slight possibility `windowGainedFocus` is never triggered if the window didn't get focus for whatever reason…

On my Windows 11 laptop, the average time (over 20 runs of the `bug4323121` test)

* for `windowGainedFocus.await`: 12 ms
* for `robot.waitForIdle`: 36 ms
* or cumulatively: 48 ms

If I change the order of calls,

* for `robot.waitForIdle`: 51 ms
* for `windowGainedFocus.await`: 0 ms
* or cumulatively: 51 ms

Both versions align well — `robot.waitForIdle` takes about 50 ms. (The version where `waitForIdle` follows `await` is slightly quicker because `await` does nothing at all whereas `waitForIdle` posts events from its own copy of the queue for handling.

Reply:
@mrserb Are any more concerns with the updated code in `JButton/bug4323121.java`?

Reply:
> @mrserb Are any more concerns with the updated code in `JButton/bug4323121.java`?

@mrserb Is there anything else you've got concerns about? If not, I'll integrate the PR.